File: sdk/typescript/src/stream.ts
/**
 * Streaming support for `daimon.provider.stream`.
 *
 * Wire shape (mirrors cmd/daimon/rpc.go::rpcStream and the server's
 * handleProviderStream):
 *
 *   1. Client sends one JSON-RPC request envelope.
 *   2. Server sends 0..N `daimon.provider.stream.delta` notifications
 *      (no `id` field per JSON-RPC 2.0 §4.1, params `{content: "..."}`).
 *   3. Server sends ONE terminal response frame carrying the request `id`
 *      and either `result` (the full `{response, injected_memory_ids?}`
 *      envelope) or `error`.
 *
 * Each frame is JSON-encoded and newline-terminated — Go's
 * json.Encoder.Encode always appends `\n`. Splitting the read buffer on
 * raw `\n` bytes is safe because JSON-encoded strings escape literal
 * newlines as `\\n`, so the only `\n` byte on the wire is the frame
 * separator.
 *
 * The connection lifecycle is different from the one-request-per-call
 * path in `./rpc.ts`: we do NOT half-close the write side after sending
 * the request because the server keeps writing until the stream
 * completes. We close fully after the terminal frame (or on early exit
 * via close()).
 */

import * as net from "node:net";

import { DaemonNotRunning, RPCError, fromErrorObject } from "./errors.js";

export const DEFAULT_STREAM_TIMEOUT_MS = 300_000;

interface Deferred<T> {
  promise: Promise<T>;
  resolve: (value: T) => void;
  reject: (reason: unknown) => void;
}

function deferred<T>(): Deferred<T> {
  let resolve!: (value: T) => void;
  let reject!: (reason: unknown) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}

type StreamEvent =
  | { kind: "delta"; content: string }
  | { kind: "terminal"; envelope: Record<string, unknown> | null }
  | { kind: "error"; err: Error };

/**
 * Async-iterable of delta strings, with the terminal envelope on `.final`.
 *
 * Usage:
 *
 *   const stream = client.provider.stream({
 *     provider: "ollama",
 *     model: "llama3.2:latest",
 *     messages: [{ role: "user", content: "say hi" }],
 *   });
 *   for await (const delta of stream) {
 *     process.stdout.write(delta);
 *   }
 *   // After iteration: stream.final is the wrapping envelope
 *   // {response: {...}, injected_memory_ids?: [...]}
 *   const usage = (stream.final?.response as { usage: unknown }).usage;
 *
 * Calling close() tears down the socket; safe to call from a finally
 * block to guarantee resource release on early exit.
 */
export class StreamHandle implements AsyncIterable<string>, AsyncIterator<string> {
  final: Record<string, unknown> | null = null;

  private readonly events: StreamEvent[] = [];
  private waiter: Deferred<void> | null = null;
  private done = false;
  private buf = "";
  private readonly sock: net.Socket;

  constructor(sock: net.Socket) {
    this.sock = sock;

    sock.on("data", (chunk: Buffer) => this.onData(chunk));
    sock.on("end", () => this.onEnd());
    sock.on("error", (err: Error) => this.onSocketError(err));
    sock.on("close", () => {
      // If the socket closes without 'end' having been received (e.g.
      // peer killed the connection), surface that as EOF so iteration
      // can terminate. onEnd is idempotent via the done flag.
      this.onEnd();
    });
  }

  [Symbol.asyncIterator](): AsyncIterator<string> {
    return this;
  }

  async next(): Promise<IteratorResult<string>> {
    while (true) {
      const event = this.events.shift();
      if (event !== undefined) {
        if (event.kind === "delta") {
          return { value: event.content, done: false };
        }
        if (event.kind === "terminal") {
          this.final = event.envelope;
          this.cleanup();
          return { value: undefined, done: true };
        }
        // error
        this.cleanup();
        throw event.err;
      }
      if (this.done) {
        // Peer closed without sending a terminal response.
        this.cleanup();
        throw new RPCError(0, "stream ended without terminal response");
      }
      // Wait for the next data event to deliver more frames.
      this.waiter = deferred<void>();
      await this.waiter.promise;
      this.waiter = null;
    }
  }

  /** Implement AsyncIterator.return() so for-await break/return tears down the socket. */
  async return(): Promise<IteratorResult<string>> {
    this.close();
    return { value: undefined, done: true };
  }

  /** Tear down the underlying socket. Idempotent. */
  close(): void {
    if (!this.sock.destroyed) {
      this.sock.destroy();
    }
    this.cleanup();
  }

  // --- internal -----------------------------------------------------------

  private cleanup(): void {
    this.done = true;
    if (this.waiter) {
      const w = this.waiter;
      this.waiter = null;
      w.resolve();
    }
  }

  private wake(): void {
    if (this.waiter) {
      const w = this.waiter;
      this.waiter = null;
      w.resolve();
    }
  }

  private onData(chunk: Buffer): void {
    this.buf += chunk.toString("utf8");
    let nl = this.buf.indexOf("\n");
    while (nl >= 0) {
      const line = this.buf.slice(0, nl);
      this.buf = this.buf.slice(nl + 1);
      if (line.trim().length > 0) {
        this.handleFrame(line);
      }
      nl = this.buf.indexOf("\n");
    }
    this.wake();
  }

  private onEnd(): void {
    if (this.done) return;
    // Drain any trailing unterminated frame, mirroring the Python reader.
    const tail = this.buf.trim();
    if (tail.length > 0) {
      this.handleFrame(tail);
      this.buf = "";
    }
    this.done = true;
    this.wake();
  }

  private onSocketError(err: Error): void {
    if (this.done) return;
    this.events.push({ kind: "error", err });
    this.done = true;
    this.wake();
  }

  private handleFrame(line: string): void {
    let frame: Record<string, unknown>;
    try {
      const parsed: unknown = JSON.parse(line);
      if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
        return;
      }
      frame = parsed as Record<string, unknown>;
    } catch (e) {
      this.events.push({
        kind: "error",
        err: new RPCError(0, `malformed stream frame: ${(e as Error).message}`),
      });
      return;
    }

    // Notification frame: no `id`, has a `method`. JSON-RPC 2.0 §4.1.
    if (!("id" in frame) && typeof frame["method"] === "string") {
      const method = frame["method"] as string;
      if (method === "daimon.provider.stream.delta") {
        const params = frame["params"];
        if (params !== null && typeof params === "object") {
          const content = (params as { content?: unknown }).content;
          if (typeof content === "string" && content.length > 0) {
            this.events.push({ kind: "delta", content });
          }
        }
      }
      // Unknown notification kinds (future tool-call deltas, role markers,
      // etc.) — ignore, forward-compat with the daemon.
      return;
    }

    // Terminal frame.
    if (frame["error"] !== undefined && frame["error"] !== null) {
      const err = fromErrorObject(frame["error"] as { code?: number; message?: string; data?: unknown });
      this.events.push({ kind: "error", err });
      return;
    }
    const result = frame["result"];
    this.events.push({
      kind: "terminal",
      envelope: (result as Record<string, unknown> | null) ?? null,
    });
  }
}

export interface OpenStreamOptions {
  timeoutMs?: number;
}

/**
 * Open a streaming RPC and return a `StreamHandle`.
 *
 * The handle owns the socket — the caller iterates (or calls close())
 * and the socket is destroyed on terminal frame, on error, or on
 * close().
 */
export function openStream(
  socketPath: string,
  method: string,
  params: unknown,
  options: OpenStreamOptions = {},
): Promise<StreamHandle> {
  const timeoutMs = options.timeoutMs ?? DEFAULT_STREAM_TIMEOUT_MS;

  return new Promise((resolve, reject) => {
    const sock = net.createConnection({ path: socketPath });
    sock.setNoDelay(true);
    let settled = false;
    const settle = (fn: () => void) => {
      if (settled) return;
      settled = true;
      fn();
    };

    sock.setTimeout(timeoutMs);
    sock.once("timeout", () => {
      settle(() => {
        sock.destroy();
        reject(new Error(`stream open timed out after ${timeoutMs}ms`));
      });
    });

    sock.once("error", (err: NodeJS.ErrnoException) => {
      const code = err.code;
      settle(() => {
        sock.destroy();
        if (code === "ENOENT" || code === "ECONNREFUSED") {
          reject(
            new DaemonNotRunning(
              `daemon ${
                code === "ENOENT" ? "socket not present" : "not accepting connections"
              } at ${socketPath} — run \`daimon unlock\` first`,
            ),
          );
          return;
        }
        reject(err);
      });
    });

    sock.once("connect", () => {
      const request: Record<string, unknown> = { jsonrpc: "2.0", method, id: 1 };
      if (params !== undefined) request["params"] = params;
      const payload = JSON.stringify(request) + "\n";
      // Critical: do NOT half-close the write side. The server reads our
      // single request, then writes notifications and the terminal frame
      // on the same connection. Use write(), not end().
      sock.write(payload, (err) => {
        if (err) {
          settle(() => {
            sock.destroy();
            reject(err);
          });
          return;
        }
        settle(() => {
          // Clear the connect-phase timeout; iteration has its own
          // lifecycle and may legitimately be idle between deltas.
          sock.setTimeout(0);
          resolve(new StreamHandle(sock));
        });
      });
    });
  });
}

File: sdk/python/daimon/_stream.py
"""Streaming support for ``daimon.provider.stream``.

Wire shape (mirrors ``cmd/daimon/rpc.go::rpcStream`` and the server's
``handleProviderStream``):

  1. Client sends one JSON-RPC request envelope.
  2. Server sends 0..N ``daimon.provider.stream.delta`` notifications
     (no ``id`` field per JSON-RPC 2.0 §4.1, params ``{content: "..."}``).
  3. Server sends ONE terminal response frame carrying the request ``id``
     and either ``result`` (the full ``{response, injected_memory_ids?}``
     envelope) or ``error``.

Each frame is JSON-encoded and newline-terminated — Go's
``json.Encoder.Encode`` always appends ``\\n``. Splitting the read buffer
on raw ``\\n`` bytes is safe because JSON strings escape literal newlines
as ``\\\\n`` (two ASCII chars), so the only ``\\n`` byte on the wire is
the frame separator.

The connection lifecycle is different from the one-request-per-call
path in :mod:`._rpc`: we do NOT half-close the write side after sending
the request because the server keeps writing until the stream
completes. We close fully after the terminal frame (or on early exit
via the context-manager protocol).
"""

from __future__ import annotations

import errno
import json
import socket
from pathlib import Path
from typing import Any, Iterator

from .errors import DaemonNotRunning, RPCError, _from_error_object

DEFAULT_TIMEOUT = 300.0


class StreamHandle:
    """Iterable of delta strings, with the terminal envelope on ``.final``.

    Usage::

        stream = client.provider.stream(
            provider="ollama",
            model="llama3.2:latest",
            messages=[{"role": "user", "content": "say hi"}],
        )
        for delta in stream:
            print(delta, end="", flush=True)
        print()
        # After iteration: stream.final is the wrapping envelope
        # {response: {...}, injected_memory_ids?: [...]}
        usage = stream.final["response"]["usage"]

    Or as a context manager — closes the socket on exit even if the
    caller breaks out of iteration early::

        with client.provider.stream(...) as stream:
            for delta in stream:
                if some_cancel_condition():
                    break
    """

    final: dict | None = None

    def __init__(self, sock: socket.socket) -> None:
        self._sock: socket.socket | None = sock
        self._buf = bytearray()
        self._done = False

    def __iter__(self) -> "StreamHandle":
        return self

    def __next__(self) -> str:
        if self._done:
            raise StopIteration
        while True:
            frame = self._read_one_frame()
            if frame is None:
                # Peer closed without sending a terminal response.
                self._finish()
                raise RPCError(0, "stream ended without terminal response")

            # Notification frame: no `id`, has a `method`. JSON-RPC 2.0 §4.1.
            if "id" not in frame and frame.get("method"):
                method = frame.get("method", "")
                if method == "daimon.provider.stream.delta":
                    params = frame.get("params") or {}
                    content = params.get("content", "") if isinstance(params, dict) else ""
                    if content:
                        return content
                # Unknown notification kinds (future tool-call deltas, role
                # markers, etc.) — ignore, forward-compat with the daemon.
                continue

            # Terminal frame.
            if frame.get("error"):
                self._finish()
                raise _from_error_object(frame["error"])
            self.final = frame.get("result")
            self._finish()
            raise StopIteration

    def __enter__(self) -> "StreamHandle":
        return self

    def __exit__(self, *_a: Any) -> None:
        self.close()

    def close(self) -> None:
        """Tear down the underlying socket. Idempotent."""
        self._done = True
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    # --- internal -----------------------------------------------------------

    def _finish(self) -> None:
        self._done = True
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    def _read_one_frame(self) -> dict | None:
        """Return the next decoded JSON object on the wire, or None at EOF."""
        if self._sock is None:
            return None
        while True:
            i = self._buf.find(b"\n")
            if i >= 0:
                line = bytes(self._buf[:i])
                del self._buf[: i + 1]
                if line.strip():
                    return json.loads(line)
                continue
            try:
                chunk = self._sock.recv(4096)
            except OSError:
                # Treat read errors as EOF — match the non-streaming
                # client's tolerant drain in :mod:`._rpc`.
                chunk = b""
            if not chunk:
                # Final trailing line without a terminator.
                if self._buf.strip():
                    line = bytes(self._buf)
                    self._buf.clear()
                    return json.loads(line)
                return None
            self._buf.extend(chunk)


def open_stream(
    socket_path: Path | str,
    method: str,
    params: Any,
    timeout: float | None = None,
) -> StreamHandle:
    """Open a streaming RPC and return a :class:`StreamHandle`.

    The handle owns the socket — the caller iterates it (or uses it as a
    context manager) and the socket is closed on terminal frame, on error,
    or on context-manager exit.
    """
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.settimeout(timeout if timeout is not None else DEFAULT_TIMEOUT)
    try:
        sock.connect(str(socket_path))
    except FileNotFoundError as e:
        sock.close()
        raise DaemonNotRunning(
            f"daemon socket not present at {socket_path} — run `daimon unlock` first"
        ) from e
    except OSError as e:
        sock.close()
        if e.errno in (errno.ECONNREFUSED, errno.ENOENT):
            raise DaemonNotRunning(
                f"daemon not accepting connections at {socket_path} — "
                f"run `daimon unlock` first"
            ) from e
        raise

    request = {
        "jsonrpc": "2.0",
        "method": method,
        **({"params": params} if params is not None else {}),
        "id": 1,
    }
    payload = (json.dumps(request, separators=(",", ":")) + "\n").encode("utf-8")
    try:
        sock.sendall(payload)
    except OSError:
        sock.close()
        raise

    # Critical: do NOT half-close the write side. The server reads our
    # single request, then writes notifications and the terminal frame on
    # the same connection — and on Linux, FIN on the read side can be
    # surfaced as a write-side error on some kernels. The non-streaming
    # half-close trick is correct for one-shot but wrong for streams.
    return StreamHandle(sock)
